protocols/kad/src/addresses.rs

```diff
@@ -22,67 +22,77 @@ use libp2p_core::Multiaddr;
 use smallvec::SmallVec;
 use std::fmt;
 
-/// List of addresses of a peer.
+/// A non-empty list of (unique) addresses of a peer in the routing table.
 #[derive(Clone)]
 pub struct Addresses {
     addrs: SmallVec<[Multiaddr; 6]>,
 }
 
 impl Addresses {
     /// Creates a new list of addresses.
-    pub fn new() -> Addresses {
-        Addresses {
-            addrs: SmallVec::new(),
-        }
+    pub fn new(addr: Multiaddr) -> Addresses {
+        let mut addrs = SmallVec::new();
+        addrs.push(addr);
+        Addresses { addrs }
+    }
+
+    /// Gets a reference to the first address in the list.
+    pub fn first(&self) -> &Multiaddr {
+        &self.addrs[0]
     }
 
-    /// Returns an iterator over the list of addresses.
+    /// Returns an iterator over the addresses.
     pub fn iter(&self) -> impl Iterator<Item = &Multiaddr> {
         self.addrs.iter()
     }
 
+    /// Returns the number of addresses in the list.
+    pub fn len(&self) -> usize {
+        self.addrs.len()
+    }
+
     /// Converts the addresses into a `Vec`.
     pub fn into_vec(self) -> Vec<Multiaddr> {
         self.addrs.into_vec()
     }
 
-    /// Returns true if the list of addresses is empty.
-    pub fn is_empty(&self) -> bool {
-        self.addrs.is_empty()
-    }
+    /// Removes the given address from the list.
+    ///
+    /// Returns true if the address was found and removed, false otherwise.
+    /// The last remaining address in the list cannot be removed.
+    ///
+    /// An address should only be removed if is determined to be invalid or
+    /// otherwise unreachable.
+    pub fn remove(&mut self, addr: &Multiaddr) -> bool {
```

Comment:
What about returning `-> Result<(), ()>` instead? IMO it's more explicit that `Ok` means that the removal was successful.

Reply:
Sounds good to me.